File: practices/interview-bit/stacks-and-queues/min-stack.py
class MinStack:
    stack = []
    minimums = []

    # @param x, an integer
    # @return an integer
    def push(self, x):
        self.stack.append(x)

        if len(self.minimums) == 0 or self.minimums[-1] >= x:
            self.minimums.append(x)

    # @return nothing
    def pop(self):
        if len(self.stack) == 0:
            return

        top = self.stack[-1]
        self.stack.pop()

        if top == self.minimums[-1]:
            self.minimums.pop()

    # @return an integer
    def top(self):
        if len(self.stack) != 0:
            return self.stack[-1]
        return -1

    # @return an integer
    def getMin(self):
        if len(self.minimums) != 0:
            return self.minimums[-1]
        return -1

    def __init__(self):
        self.stack = []
        self.minimums = []

File: practices/interview-bit/stacks-and-queues/min-stack.cpp
// Solution 1: Maintain 2 stacks
// O(N + X) in space, where X = number of places where minimum changes or the element is the same as the minimum
deque<int> s;
deque<int> sMin;

MinStack::MinStack() {
    s.clear();
    sMin.clear();
}

void MinStack::push(int x) {
    s.push_back(x);

    if (sMin.empty() || sMin.back() >= x) {
        sMin.push_back(x);
    }
}

void MinStack::pop() {
    if (s.empty()) return;

    int top = s.back();
    s.pop_back();

    if (top == sMin.back()) {
        sMin.pop_back();
    }
}

int MinStack::top() {
    if (!s.empty()) return s.back();
    return -1;
}

int MinStack::getMin() {
    if (!sMin.empty()) return sMin.back();
    return -1;
}
